Index: backports-4.19.237-1/net/mac80211/key.c
===================================================================
--- backports-4.19.237-1.orig/net/mac80211/key.c
+++ backports-4.19.237-1/net/mac80211/key.c
@@ -390,6 +390,10 @@ ieee80211_key_alloc(u32 cipher, int idx,
 {
 	struct ieee80211_key *key;
 	int i, j, err;
+	union {
+		u64 val64;
+		u8 val8[8];
+	} tx_pn;
 
 	if (WARN_ON(idx < 0 || idx >= NUM_DEFAULT_KEYS + NUM_DEFAULT_MGMT_KEYS))
 		return ERR_PTR(-EINVAL);
@@ -435,6 +439,13 @@ ieee80211_key_alloc(u32 cipher, int idx,
 				for (j = 0; j < IEEE80211_CCMP_PN_LEN; j++)
 					key->u.ccmp.rx_pn[i][j] =
 						seq[IEEE80211_CCMP_PN_LEN - j - 1];
+			/* If tx_pn provided after rx_pn, set it */
+			if (seq_len >= (IEEE80211_CCMP_PN_LEN * 2)) {
+				printk("GLENNON: Setting tx_pn\n");
+				for (i = 0; i < IEEE80211_CCMP_PN_LEN; i++)
+					tx_pn.val8[i] = seq[IEEE80211_CCMP_PN_LEN + i];
+				atomic64_set(&(key->conf.tx_pn), tx_pn.val64);
+			}
 		}
 		/*
 		 * Initialize AES key state here as an optimization so that
@@ -455,6 +466,13 @@ ieee80211_key_alloc(u32 cipher, int idx,
 			for (j = 0; j < IEEE80211_CCMP_256_PN_LEN; j++)
 				key->u.ccmp.rx_pn[i][j] =
 					seq[IEEE80211_CCMP_256_PN_LEN - j - 1];
+		/* If tx_pn provided after rx_pn, set it */
+		if (seq && seq_len >= (IEEE80211_CCMP_256_PN_LEN * 2)) {
+			printk("GLENNON: Setting tx_pn\n");
+			for (i = 0; i < IEEE80211_CCMP_256_PN_LEN; i++)
+				tx_pn.val8[i] = seq[IEEE80211_CCMP_256_PN_LEN + i];
+			atomic64_set(&(key->conf.tx_pn), tx_pn.val64);
+		}
 		/* Initialize AES key state here as an optimization so that
 		 * it does not need to be initialized for every packet.
 		 */
@@ -477,6 +495,13 @@ ieee80211_key_alloc(u32 cipher, int idx,
 			for (j = 0; j < IEEE80211_CMAC_PN_LEN; j++)
 				key->u.aes_cmac.rx_pn[j] =
 					seq[IEEE80211_CMAC_PN_LEN - j - 1];
+		/* If tx_pn provided after rx_pn, set it */
+		if (seq && seq_len >= (IEEE80211_CMAC_PN_LEN * 2)) {
+			printk("GLENNON: Setting tx_pn\n");
+			for (i = 0; i < IEEE80211_CMAC_PN_LEN; i++)
+				tx_pn.val8[i] = seq[IEEE80211_CMAC_PN_LEN + i];
+			atomic64_set(&(key->conf.tx_pn), tx_pn.val64);
+		}
 		/*
 		 * Initialize AES key state here as an optimization so that
 		 * it does not need to be initialized for every packet.
@@ -497,6 +522,13 @@ ieee80211_key_alloc(u32 cipher, int idx,
 			for (j = 0; j < IEEE80211_GMAC_PN_LEN; j++)
 				key->u.aes_gmac.rx_pn[j] =
 					seq[IEEE80211_GMAC_PN_LEN - j - 1];
+		/* If tx_pn provided after rx_pn, set it */
+		if (seq && seq_len >= (IEEE80211_GMAC_PN_LEN * 2)) {
+			printk("GLENNON: Setting tx_pn\n");
+			for (i = 0; i < IEEE80211_GMAC_PN_LEN; i++)
+				tx_pn.val8[i] = seq[IEEE80211_GMAC_PN_LEN + i];
+			atomic64_set(&(key->conf.tx_pn), tx_pn.val64);
+		}
 		/* Initialize AES key state here as an optimization so that
 		 * it does not need to be initialized for every packet.
 		 */
@@ -516,6 +548,13 @@ ieee80211_key_alloc(u32 cipher, int idx,
 			for (j = 0; j < IEEE80211_GCMP_PN_LEN; j++)
 				key->u.gcmp.rx_pn[i][j] =
 					seq[IEEE80211_GCMP_PN_LEN - j - 1];
+		/* If tx_pn provided after rx_pn, set it */
+		if (seq && seq_len >= (IEEE80211_GCMP_PN_LEN * 2)) {
+			printk("GLENNON: Setting tx_pn\n");
+			for (i = 0; i < IEEE80211_GCMP_PN_LEN; i++)
+				tx_pn.val8[i] = seq[IEEE80211_GCMP_PN_LEN + i];
+			atomic64_set(&(key->conf.tx_pn), tx_pn.val64);
+		}
 		/* Initialize AES key state here as an optimization so that
 		 * it does not need to be initialized for every packet.
 		 */
@@ -537,7 +576,7 @@ ieee80211_key_alloc(u32 cipher, int idx,
 			key->conf.iv_len = cs->hdr_len;
 			key->conf.icv_len = cs->mic_len;
 			for (i = 0; i < IEEE80211_NUM_TIDS + 1; i++)
-				for (j = 0; j < seq_len; j++)
+				for (j = 0; j < min(seq_len, (size_t)IEEE80211_CCMP_PN_LEN); j++)
 					key->u.gen.rx_pn[i][j] =
 							seq[seq_len - j - 1];
 			key->flags |= KEY_FLAG_CIPHER_SCHEME;
Index: backports-4.19.237-1/net/wireless/util.c
===================================================================
--- backports-4.19.237-1.orig/net/wireless/util.c
+++ backports-4.19.237-1/net/wireless/util.c
@@ -368,8 +368,10 @@ int cfg80211_validate_key_settings(struc
 		case WLAN_CIPHER_SUITE_BIP_CMAC_256:
 		case WLAN_CIPHER_SUITE_BIP_GMAC_128:
 		case WLAN_CIPHER_SUITE_BIP_GMAC_256:
-			if (params->seq_len != 6)
+			if (params->seq_len < 6) {
+				pr_info("validate-key:  Invalid seq-len: %d < 6\n", params->seq_len);
 				return -EINVAL;
+			}
 			break;
 		}
 	}
