--- a/hostapd/Makefile
+++ b/hostapd/Makefile
@@ -179,6 +179,11 @@ LIBS_h += -lgcov
 LIBS_n += -lgcov
 endif
 
+ifndef WPA_IGNORE_CONFIG_ERRORS
+# define WPA_IGNORE_CONFIG_ERRORS to ignore possible wrong network configurations
+CFLAGS += -DWPA_IGNORE_CONFIG_ERRORS
+endif
+
 ifndef CONFIG_NO_DUMP_STATE
 # define HOSTAPD_DUMP_STATE to include support for dumping internal state
 # through control interface commands (undefine it, if you want to save in
